### ISSUE TITLE
[Quality] Add Input Validation for Public APIs

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2642]
## Problem

Public API functions may not validate inputs, leading to cryptic errors or undefined behavior.

## Validation Needed

**Shape validation:**
```mojo
fn matmul(a: ExTensor, b: ExTensor) -> ExTensor:
    # Add validation
    if a.dim() != 2 or b.dim() != 2:
        raise Error("matmul requires 2D tensors, got shapes " + str(a.shape()) + " and " + str(b.shape()))
    
    if a.shape()[1] != b.shape()[0]:
        raise Error("Inner dimensions must match: " + str(a.shape()[1]) + " != " + str(b.shape()[0]))
    
    # ... implementation
```

**Range validation:**
```mojo
fn slice(tensor: ExTensor, axis: Int, start: Int, end: Int) -> ExTensor:
    # Validate axis
    if axis < 0 or axis >= tensor.dim():
        raise Error("Axis " + str(axis) + " out of range for " + str(tensor.dim()) + "D tensor")
    
    # Validate range
    if start < 0 or end > tensor.shape()[axis]:
        raise Error("Slice range [" + str(start) + ":" + str(end) + "] out of bounds for dimension size " + str(tensor.shape()[axis]))
    
    # ... implementation
```

**DType validation:**
```mojo
fn requires_float_dtype(tensor: ExTensor):
    if tensor.dtype() not in [DType.float16, DType.float32, DType.float64]:
        raise Error("Operation requires float dtype, got " + str(tensor.dtype()))
```

## Principles

1. **Validate at API boundaries** (public functions)
2. **Fail fast** with clear messages
3. **Don't validate internal functions** (performance)
4. **Use assertions** for internal checks (debug mode only)

## Impact

- **Severity**: LOW
- **Robustness**: Catch errors early
- **Debugging**: Clear error messages at API boundary

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a complete picture of the codebase. Let me create a comprehensive implementation plan.

---

# Implementation Plan: Issue #2642 - Add Input Validation for Public APIs

## Summary

This issue addresses the need for systematic input validation across all public API functions in the ML Odyssey tensor library. The codebase already has a `validation.mojo` module with reusable validation functions, but these are only used in 3 of 52+ core files. The goal is to extend validation coverage to all public APIs following the principles: **validate at API boundaries**, **fail fast with clear messages**, and **don't validate internal functions** (for performance).

### Current State Analysis

**Existing Validation Infrastructure:**
- `shared/core/validation.mojo` provides 5 reusable validators: `validate_tensor_shape`, `validate_tensor_dtype`, `validate_matching_tensors`, `validate_2d_input`, `validate_4d_input`
- Currently only used in: `layers/conv2d.mojo`, `layers/batchnorm.mojo`, `layers/linear.mojo`

**Files with Partial/Inline Validation:**
- `matmul.mojo` - Has inline dtype/2D/dimension checks
- `shape.mojo` - Has inline element count validation
- `pooling.mojo` - Has inline 4D input validation
- `conv.mojo` - Has inline 4D input/kernel validation
- `extensor.mojo` - Has memory size and bounds validation

**Files with Missing/Minimal Validation:**
- `arithmetic.mojo` - Only dtype match check (no shape broadcast validation)
- `reduction.mojo` - No input validation
- `activation.mojo` - No input validation
- `comparison.mojo` - No input validation
- `elementwise.mojo` - No input validation
- `loss.mojo` - Limited validation

---

## Step-by-Step Implementation Tasks

### Phase 1: Extend Validation Module (Priority: High)

**Task 1.1: Add New Validation Functions to `validation.mojo`**

Add the following new validators to `shared/core/validation.mojo`:

```mojo
# 1. Axis range validation
fn validate_axis(tensor: ExTensor, axis: Int, name: String) raises

# 2. Slice range validation  
fn validate_slice_range(tensor: ExTensor, axis: Int, start: Int, end: Int, name: String) raises

# 3. Float dtype requirement (for ops like softmax, sigmoid)
fn validate_float_dtype(tensor: ExTensor, name: String) raises

# 4. Positive shape dimensions
fn validate_positive_shape(shape: List[Int], name: String) raises

# 5. Matmul dimension compatibility
fn validate_matmul_dims(a: ExTensor, b: ExTensor, a_name: String, b_name: String) raises

# 6. Broadcasting compatibility check
fn validate_broadcast_compatible(a: ExTensor, b: ExTensor, a_name: String, b_name: String) raises

# 7. 1D tensor validation
fn validate_1d_input(tensor: ExTensor, name: String) raises

# 8. 3D tensor validation
fn validate_3d_input(tensor: ExTensor, name: String) raises

# 9. Non-empty tensor validation
fn validate_non_empty(tensor: ExTensor, name: String) raises

# 10. Matching dtype validation (for binary ops)
fn validate_matching_dtype(a: ExTensor, b: ExTensor, a_name: String, b_name: String) raises
```

**Rationale:** These cover all validation patterns needed across the codebase per the issue requirements.

---

### Phase 2: Matrix Operations Validation (Priority: High)

**Task 2.1: Update `matmul.mojo` to Use Centralized Validators**

Replace inline validation with `validation.mojo` functions:

```mojo
# Current inline validation (lines 82-101):
if a.dtype() != b.dtype() or a.dtype() != c.dtype():
    raise Error("matmul_typed: all tensors must have the same dtype")
# ...

# Replace with:
validate_matching_dtype(a, b, "a", "b")
validate_matching_dtype(a, c, "a", "c")
validate_2d_input(a, "a")
validate_2d_input(b, "b")
validate_matmul_dims(a, b, "a", "b")
```

**Files to modify:**
- `/home/mvillmow/ml-odyssey/shared/core/matmul.mojo` (functions: `matmul_typed`, `matmul_simd`, `matmul_tiled`, `matmul_transpose`, `matmul`, `matmul_optimized`)

**Task 2.2: Update `matrix.mojo` Validation**

Standardize validation in:
- `matmul()` - lines 464-530
- `transpose()` - lines 628-731 (add axis validation)
- `dot()` - lines 732-819 (add 1D validation)

---

### Phase 3: Arithmetic Operations Validation (Priority: High)

**Task 3.1: Add Broadcasting Validation to `arithmetic.mojo`**

Update `_dispatch_broadcast_binary()` to validate broadcast compatibility before operations:

```mojo
fn _dispatch_broadcast_binary[...](a: ExTensor, b: ExTensor) raises -> ExTensor:
    # Validate dtypes match
    validate_matching_dtype(a, b, "a", "b")
    
    # Validate shapes are broadcast-compatible
    validate_broadcast_compatible(a, b, "a", "b")
    
    # ... rest of implementation
```

**Functions to update:**
- `add()`, `subtract()`, `multiply()`, `divide()`, `floor_divide()`, `modulo()`, `power()`
- Backward functions: `add_backward()`, `subtract_backward()`, etc.

---

### Phase 4: Activation Functions Validation (Priority: Medium)

**Task 4.1: Add Validation to Float-Required Activations**

Functions requiring float dtype validation:
- `sigmoid()` - requires float dtype
- `tanh()` - requires float dtype
- `softmax()` - requires float dtype + axis validation
- `gelu()` - requires float dtype
- `swish()` - requires float dtype

Example for `softmax()`:
```mojo
fn softmax(tensor: ExTensor, axis: Int = -1) raises -> ExTensor:
    validate_float_dtype(tensor, "input")
    validate_axis(tensor, axis, "axis")
    # ... implementation
```

**Task 4.2: Add Validation to ReLU Family**

Functions with range parameter validation:
- `leaky_relu()` - validate alpha is reasonable (0 < alpha < 1)
- `elu()` - validate alpha > 0
- `selu()` - no additional validation needed

---

### Phase 5: Reduction Operations Validation (Priority: Medium)

**Task 5.1: Add Axis Validation to Reduction Functions**

Update `shared/core/reduction.mojo`:
- `sum()` - validate axis in range if provided
- `mean()` - validate axis in range if provided
- `max_reduce()` - validate axis in range if provided
- `min_reduce()` - validate axis in range if provided

```mojo
fn sum(tensor: ExTensor, axis: Int = -1, keepdims: Bool = False) raises -> ExTensor:
    validate_non_empty(tensor, "input")
    if axis != -1:  # -1 means reduce all
        validate_axis(tensor, axis, "axis")
    # ... implementation
```

---

### Phase 6: Shape Operations Validation (Priority: Medium)

**Task 6.1: Consolidate Inline Validation in `shape.mojo`**

Replace inline validation with centralized functions:
- `reshape()` - currently has inline validation
- `view()` - currently has inline validation
- `squeeze()` - add explicit dimension validation
- `unsqueeze()` - add explicit dimension validation
- `concatenate()` - standardize error messages
- `stack()` - standardize error messages

---

### Phase 7: Pooling and Convolution Validation (Priority: Medium)

**Task 7.1: Enhance `pooling.mojo` Validation**

Add validation for:
- Kernel size > 0
- Stride > 0 (when not using default)
- Padding >= 0
- Output dimensions are positive

```mojo
fn maxpool2d(...) raises -> ExTensor:
    validate_4d_input(x, "input")
    if kernel_size <= 0:
        raise Error("maxpool2d: kernel_size must be positive, got " + String(kernel_size))
    if stride < 0:
        raise Error("maxpool2d: stride must be non-negative, got " + String(stride))
    if padding < 0:
        raise Error("maxpool2d: padding must be non-negative, got " + String(padding))
    # ... implementation
```

**Task 7.2: Enhance `conv.mojo` Validation**

Standardize validation across:
- `conv2d()` - already has partial validation
- `conv2d_no_bias()` - needs same validation
- `conv2d_backward()` - gradient shape validation

---

### Phase 8: Loss Functions Validation (Priority: Low)

**Task 8.1: Add Validation to Loss Functions**

Update `shared/core/loss.mojo`:
- `cross_entropy()` - validate shapes match, validate logits are 2D
- `cross_entropy_backward()` - validate gradient shapes

Update `shared/core/loss_utils.mojo`:
- `mse_loss()` - validate matching shapes
- `l1_loss()` - validate matching shapes
- `nll_loss()` - validate log_probs is 2D, targets is 1D

---

### Phase 9: Comparison Operations Validation (Priority: Low)

**Task 9.1: Add Validation to Comparison Functions**

Update `shared/core/comparison.mojo`:
- `equal()`, `not_equal()`, `less()`, `less_equal()`, `greater()`, `greater_equal()`
- All need: `validate_matching_dtype()` and `validate_broadcast_compatible()`

---

## Files to Modify/Create

### Files to Modify (in priority order):

| File | Changes | Priority |
|------|---------|----------|
| `shared/core/validation.mojo` | Add 10 new validation functions | High |
| `shared/core/matmul.mojo` | Replace inline with centralized validation | High |
| `shared/core/matrix.mojo` | Add/standardize validation | High |
| `shared/core/arithmetic.mojo` | Add broadcast validation | High |
| `shared/core/activation.mojo` | Add float dtype validation | Medium |
| `shared/core/reduction.mojo` | Add axis validation | Medium |
| `shared/core/shape.mojo` | Consolidate validation | Medium |
| `shared/core/pooling.mojo` | Add parameter validation | Medium |
| `shared/core/conv.mojo` | Standardize validation | Medium |
| `shared/core/loss.mojo` | Add shape validation | Low |
| `shared/core/loss_utils.mojo` | Add shape validation | Low |
| `shared/core/comparison.mojo` | Add dtype/broadcast validation | Low |
| `shared/core/elementwise.mojo` | Add float dtype validation | Low |

### Files to Create:

| File | Purpose |
|------|---------|
| `tests/shared/core/test_validation_extended.mojo` | Tests for new validation functions |
| `tests/shared/core/test_api_validation.mojo` | Integration tests for API boundary validation |

---

## Testing Approach

### Unit Tests for New Validators (Phase 1)

For each new validation function in `validation.mojo`, add tests following the existing pattern in `test_validation.mojo`:

```mojo
# Example test structure for validate_axis
fn test_validate_axis_valid() raises:
    var x = zeros([3, 4, 5], DType.float32)
    validate_axis(x, 0, "x")  # First axis
    validate_axis(x, 1, "x")  # Second axis
    validate_axis(x, 2, "x")  # Third axis
    validate_axis(x, -1, "x") # Last axis (negative indexing)

fn test_validate_axis_invalid() raises:
    var x = zeros([3, 4, 5], DType.float32)
    var error_raised = False
    try:
        validate_axis(x, 5, "input")  # Out of range
    except e:
        error_raised = True
        var error_msg = String(e)
        assert_true("Axis 5 out of range" in error_msg, "Error should mention axis out of range")
    assert_true(error_raised, "Error should be raised for invalid axis")
```

### Integration Tests for API Boundaries

Test that each public API function properly rejects invalid inputs:

```mojo
# Test matmul rejects non-2D tensors
fn test_matmul_rejects_3d_tensor() raises:
    var a = zeros([2, 3, 4], DType.float32)  # 3D
    var b = zeros([4, 5], DType.float32)     # 2D
    var error_raised = False
    try:
        _ = matmul(a, b)
    except e:
        error_raised = True
        assert_true("2D" in String(e), "Error should mention 2D requirement")
    assert_true(error_raised, "matmul should reject 3D tensors")

# Test softmax rejects int dtype
fn test_softmax_rejects_int_dtype() raises:
    var x = zeros([3, 4], DType.int32)
    var error_raised = False
    try:
        _ = softmax(x)
    except e:
        error_raised = True
        assert_true("float" in String(e), "Error should mention float requirement")
    assert_true(error_raised, "softmax should reject int dtype")
```

### Error Message Quality Tests

Verify error messages include:
- Operation name (e.g., "matmul:", "softmax:")
- What was expected vs what was received
- Tensor names for clarity

```mojo
fn test_error_message_quality() raises:
    var a = zeros([3, 4], DType.float32)
    var b = zeros([5, 6], DType.float64)  # Different dtype
    
    try:
        _ = add(a, b)
    except e:
        var msg = String(e)
        # Error should include:
        assert_true("add" in msg or "dtype" in msg, "Should mention operation or issue")
        assert_true("float32" in msg and "float64" in msg, "Should mention both dtypes")
```

---

## Success Criteria

### Functional Requirements

1. **All public API functions validate inputs at entry** - No invalid input reaches internal computation
2. **Clear, actionable error messages** - Each error includes operation name, expected vs actual values
3. **Consistent validation patterns** - All similar operations use the same validation functions
4. **No performance impact on valid inputs** - Validation only at API boundaries, not internal functions

### Quality Metrics

| Metric | Target |
|--------|--------|
| Public APIs with validation | 100% of tensor operations |
| Validation coverage in tests | 100% of new validators tested |
| Error message clarity | All errors include operation name + expected vs actual |
| Code duplication | < 5 lines of repeated validation logic |

### Verification Checklist

- [ ] All 10 new validators added to `validation.mojo`
- [ ] All validators have unit tests in `test_validation_extended.mojo`
- [ ] `matmul.mojo` uses centralized validators
- [ ] `arithmetic.mojo` validates broadcast compatibility
- [ ] `activation.mojo` validates float dtype where required
- [ ] `reduction.mojo` validates axis parameters
- [ ] `pooling.mojo` validates all parameters
- [ ] `conv.mojo` validation is standardized
- [ ] Integration tests verify API boundaries
- [ ] All existing tests still pass
- [ ] Pre-commit hooks pass (formatting, linting)

---

## Implementation Notes

### Principle 1: Validate at API Boundaries Only

```mojo
# GOOD - Validate in public function
fn matmul(a: ExTensor, b: ExTensor) raises -> ExTensor:
    validate_2d_input(a, "a")
    validate_2d_input(b, "b")
    validate_matmul_dims(a, b, "a", "b")
    return _matmul_impl(a, b)  # Internal function - no validation

# BAD - Validation in internal function (performance hit)
fn _matmul_impl(a: ExTensor, b: ExTensor) raises -> ExTensor:
    validate_2d_input(a, "a")  # Unnecessary - already validated
    # ...
```

### Principle 2: Fail Fast with Clear Messages

```mojo
# GOOD - Clear, actionable error
raise Error("matmul: inner dimensions must match, got " + String(a.shape()[1]) 
            + " (a) vs " + String(b.shape()[0]) + " (b)")

# BAD - Cryptic error
raise Error("incompatible dimensions")
```

### Principle 3: Use Debug Assertions for Internal Checks

```mojo
# For internal sanity checks that should never fail if API validation is correct
fn _internal_helper(tensor: ExTensor) -> ExTensor:
    debug_assert(len(tensor.shape()) == 2, "Internal error: expected 2D tensor")
    # ... implementation
```

---

## Estimated Effort

| Phase | Tasks | Estimated Time |
|-------|-------|----------------|
| Phase 1: Extend Validation Module | 1 | 2-3 hours |
| Phase 2: Matrix Operations | 2 | 2 hours |
| Phase 3: Arithmetic Operations | 1 | 1-2 hours |
| Phase 4: Activation Functions | 2 | 2 hours |
| Phase 5: Reduction Operations | 1 | 1 hour |
| Phase 6: Shape Operations | 1 | 1-2 hours |
| Phase 7: Pooling/Convolution | 2 | 2 hours |
| Phase 8: Loss Functions | 1 | 1 hour |
| Phase 9: Comparison Operations | 1 | 1 hour |
| Testing & Verification | - | 3-4 hours |
| **Total** | **12** | **16-20 hours** |

---

## Dependencies

- No external dependencies
- All changes are internal to `shared/core/`
- No breaking API changes (only adding validation, not changing signatures)

---

## Risk Assessment

| Risk | Mitigation |
|------|------------|
| Performance regression from validation overhead | Validate only at public API entry points |
| Breaking existing tests | Run full test suite after each phase |
| Inconsistent error message formats | Create error message templates in validation.mojo |
| Missing edge cases in validation | Add comprehensive test coverage for each validator |

---

## Resource Usage

- Tool calls used: 10/50
- Steps used: 10/50
